src/mds/Locker.cc

```diff
@@ -4153,7 +4153,7 @@ void Locker::caps_tick()
       CInode *in = need_snapflush_inodes.front();
       if (in->last_dirstat_prop >= cutoff)
 	break;
-      in->item_caps.remove_myself();
+      in->item_to_flush.remove_myself();
```

Comment:
You may also need to do this in `void MDCache::remove_inode(CInode *o)`.

Reply:
I'm seeing this patch in isolation with the locker code. `MDCache::remove_inode` is called all over the MDS code, won't removing `in->item_caps.remove_myself();` result in unexpected behaviour? Maybe I can keep both `in->item_caps.remove_myself();` and `in->item_to_flush.remove_myself();` in this func?

Reply:
> Maybe I can keep both `in->item_caps.remove_myself();` and `in->item_to_flush.remove_myself();` in this func?

Since `need_snapflush_inodes` works with `item_to_flush` now, I think it makes sense to just have `in->item_to_flush.remove_myself()` here.

Reply:
I'm a bit scared since `MDCache::remove_inode` is not limited to locker code but spread across entire MDS code, with the proposed change if the item_caps is not removed then would it impact funcs like e.g. `CDir::try_remove_dentries_for_stray`?

Reply:
In MDCache::remove_inode(CInode *):

```
  // remove from inode map
  if (o->last == CEPH_NOSNAP) {
    inode_map.erase(o->ino());
  } else {
    o->item_caps.remove_myself();
    snap_inode_map.erase(o->vino());
  }
```

If it's a snapped inode, it's a part of a snap realm and `item_caps` is used, which I think is correct. We need a `->item_to_flush.remove_myself()` - question is should it be under a conditional check (is_on_list()).

Reply:
> if we take a look at `MDCache::journal_cow_dentry` - there is a call to `Locker::mark_need_snapflush_inode` only if old dentry's snap caps exist and if we check `Locker::mark_need_snapflush_inode` then only after conditional check `item_to_flush` is pushed in. `MDCache::remove_inode` is used across MDS code, it doesn't necessarily mean that when it is called there are always some snapped inodes to be flushed(?). I don't think that this sort of general func should remove an element without the conditional check.

Agreed.

Reply:
@dparmar18 Sorry for late reply.

I know the `item_to_flush` is isolated within the locker code, but when deleting the `Cinode` we need to make sure we have cleaned it from all the lists, else we can get the call trace or abort, just like this bug we hit https://tracker.ceph.com/issues/56698.

Please note all the other items in the `CInode` have done this in `MDCache::remove_inode()`.

Reply:
> @dparmar18 Sorry for late reply.
> 
> I know the `item_to_flush` is isolated within the locker code, but when deleting the `Cinode` we need to make sure we have cleaned it from all the lists, else we can get the call trace or abort, just like this bug we hit https://tracker.ceph.com/issues/56698.
> 
> Please note all the other items in the `CInode` have done this in `MDCache::remove_inode()`.

Yes, we should not leave items hanging off any list on teardown. The question here was is it safe to call o->item_to_flush.remove_myself() directly or have it under is_on_list() check.

Reply:
> 
> Yes, we should not leave items hanging off any list on teardown. The question here was is it safe to call o->item_to_flush.remove_myself() directly or have it under is_on_list() check.

Yeah, I think it is safe, please see the `remove_myself()` function in `elist`, it will do nothing when it's `empty`:

```
 49     bool remove_myself() {
 50       if (_next == this) {
 51         ceph_assert(_prev == this);
 52         return false;          
 53       }
 54       _next->_prev = _prev;    
 55       _prev->_next = _next;    
 56       _prev = _next = this;    
 57       return true;
 58     }
```

I don't think the `is_on_list()` makes any sense here:

```
 46     bool empty() const { return _prev == this; }
 47     bool is_on_list() const { return !empty(); }     
```

Once we see a assert in above `Line#51` then it should be a bug in `elist` itself.

Reply:
> Yeah, I think it is safe, please see the `remove_myself()` function in `elist`, it will do nothing when it's `empty`:

basically, the is_on_list() check if much useful when trying to push_back an item to check if its already linked, but there are places where remove_myself() is under is_on_list() conditional.

Reply:
Yeah. Some places are really doing this, but also I can see for the same `item` some other places are not. Which means the `is_on_list()` is not a must for the `remove_myself()`.

Usually when `remove_myself()` is under `is_on_list()` it will be used to do some extra work, for example to print some debug logs.

Reply:
okay, thanks for the insights on `is_on_list()` and `remove_myself()` @lxbsz, will update the PR